File: cyberpath/tmp_add_requires.py
from pathlib import Path
import re
path = Path('lib/data.js')
text = path.read_text(encoding='utf-8')
lines = text.splitlines()
out_lines = []
inside_items = False
item_lines = []
current_title = None
last_title = None
for line in lines:
    if not inside_items and re.match(r'\s*items\s*:\s*\[', line):
        inside_items = True
        out_lines.append(line)
        continue
    if inside_items:
        if item_lines or re.match(r'\s*\{\s*$', line):
            item_lines.append(line)
            title_match = re.match(r'\s*title\s*:\s*"([^"]+)"\s*,?\s*$', line)
            if title_match and current_title is None:
                current_title = title_match.group(1)
            if re.match(r'\s*\}\s*,?\s*$', line):
                item_text = '\n'.join(item_lines)
                if 'requires:' not in item_text:
                    reqs = '[]' if last_title is None else f'["{last_title}"]'
                    indent = re.match(r'(\s*)\}', item_lines[-1]).group(1)
                    req_line = f'{indent}  requires: {reqs},'
                    item_lines.insert(-1, req_line)
                out_lines.extend(item_lines)
                last_title = current_title or last_title
                item_lines = []
                current_title = None
            continue
        if re.match(r'\s*\]\s*,?\s*$', line):
            inside_items = False
            out_lines.append(line)
            continue
        out_lines.append(line)
    else:
        out_lines.append(line)
new_text = '\n'.join(out_lines) + ('\n' if text.endswith('\n') else '')
if new_text != text:
    path.write_text(new_text, encoding='utf-8')
    print('Updated')
else:
    print('No changes')

File: cyberpath/tmp_add_requires.js
const fs = require('fs');
const path = require('path');
const filePath = path.join(__dirname, 'lib', 'data.js');
let text = fs.readFileSync(filePath, 'utf8');
const lines = text.split(/\r?\n/);
let out = [];
let insideItems = false;
let itemLines = [];
let currentTitle = null;
let lastTitle = null;
for (let line of lines) {
  if (!insideItems && /^\s*items\s*:\s*\[/.test(line)) {
    insideItems = true;
    out.push(line);
    continue;
  }
  if (insideItems) {
    if (itemLines.length > 0 || /^\s*\{\s*$/.test(line)) {
      itemLines.push(line);
      const titleMatch = line.match(/^\s*title\s*:\s*"([^"]+)"\s*,?\s*$/);
      if (titleMatch && currentTitle == null) {
        currentTitle = titleMatch[1];
      }
      if (/^\s*\}\s*,?\s*$/.test(line)) {
        const itemText = itemLines.join('\n');
        if (!/\brequires\s*:\s*\[/.test(itemText)) {
          const reqs = lastTitle == null ? '[]' : JSON.stringify([lastTitle]);
          const indent = (line.match(/^(\s*)\}/) || ['','']).slice(1)[0];
          const reqLine = `${indent}  requires: ${reqs},`;
          itemLines.splice(itemLines.length - 1, 0, reqLine);
        }
        out.push(...itemLines);
        lastTitle = currentTitle || lastTitle;
        itemLines = [];
        currentTitle = null;
      }
      continue;
    }
    if (/^\s*\]\s*,?\s*$/.test(line)) {
      insideItems = false;
      out.push(line);
      continue;
    }
    out.push(line);
    continue;
  }
  out.push(line);
}
const newText = out.join('\n') + (text.endsWith('\n') ? '\n' : '');
if (newText !== text) {
  fs.writeFileSync(filePath, newText, 'utf8');
  console.log('Updated data.js');
} else {
  console.log('No changes needed');
}
